### PR TITLE
Add ghc 8.0.1

### PR DESCRIPTION
**Pull request by ncihnegn**
I am quite satisfied that this builds without an inplace _mingw_ tree and runs without that.
In other words, this is a clean _ghc_ build for _mingw_, if you know what I am talking about.
But it still needs _ghc_ along with _alex_ and _happy_ to build itself, which we can't get around.
I am not sure how to put these in _makedepends_ or _prepare()_.

It also works with _mingw-llvm_ 3.8.

### CODE REVIEW

**Comment by Alexpux** [comment]
Is it really depends on MSYS2 perl?
Also check other PKGBUILD's and do the same for GHC - header, pkgdesc, _realname, arch.
Remove wxWidgets from this pull request

**Comment by ncihnegn** [comment]
Fix committed but this won't build without ghc/alex/happy.

**Comment by Alexpux** [comment]
So first we need build it with third party compiler and then we can use our own to rebuild it ourself?

File: mingw-w64-ghc/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,87 @@
+_realname=ghc
+pkgbase=mingw-w64-${_realname}
+pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
+pkgver=8.0.1
+pkgrel=1
+pkgdesc='The Glorious Glasgow Haskell Compilation System'
```
Comment (filnet) [inline mingw-w64-ghc/PKGBUILD]: Should be "The Glasgow Haskell Compiler" ? 
I think that humor does not have its place here (unless GHC is also known as GGHCS...)

Comment (ncihnegn) [inline mingw-w64-ghc/PKGBUILD]: I think they are equal.
➜  ~ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1

**Comment by ncihnegn** [comment]
Yes, we need ghc/alex/happy to bootstrap.
Shall I upload a prebuilt binary package?
I am also working on adding alex/happy packages.

**Comment by ncihnegn** [comment]
The build is good although strip fails.
What shall I do to pass the CI checks? Add a wrapper like https://gist.github.com/michalfapso/833d4198244698e4266e?
If you think this is ready, I will change from the quickest build to performance build.

**Comment by mati865** [comment]
> What shall I do to pass the CI checks?
AppVeyor failed due to 1 hour limit, there is nothing you can do.
Tea 32 bit failed because of wrong file name
```
==> ERROR: Failure while downloading https://www.haskell.org/cabal/release/cabal-install-1.24.0.0/cabal-install-1.24.0.0-i686-unknown-mingw32.zip
```
Tea 64 bit could fail due to Wine bug but I'm not sure.

**Comment by ncihnegn** [comment]
Now tea CI works. How to restart appveyor that fails with a network problem?

File: mingw-w64-ghc/PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,103 @@
+_realname=ghc
```
Comment (mati865) [inline mingw-w64-ghc/PKGBUILD]: Add Maintainer like here: https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-PKGBUILD-templates/PKGBUILD.autotools-tarball

**Comment by mati865** [comment]
AppVeyor will restart when to push to to ghc branch.

File: mingw-w64-ghc/0001-Switch-to-LLVM-version-3.8.patch [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
++        <> char ')'
++pprSpecialStatic stat = ppr stat
++
++
+ pprStaticArith :: LlvmStatic -> LlvmStatic -> LitString -> LitString -> String -> SDoc
+ pprStaticArith s1 s2 int_op float_op op_name =
+   let ty1 = getStatType s1
+diff --git a/configure.ac b/configure.ac
+index 664deb4..f4b839e 100644
+--- a/configure.ac
```
Comment (Mistuke) [inline mingw-w64-ghc/0001-Switch-to-LLVM-version-3.8.patch]: You don't set the release flag here, which will make GHC report itself as a dev build. Or is this the intention of the package?

**Comment by mati865** [comment]
What is the exact problem with mingw compilers from MSYS2?

**Comment by Mistuke** [comment]
@mati865 I don't quite understand the question :)

**Comment by Mistuke** [comment]
@mati865 Historically, we've always used the mingw compilers with GHC, and now mingw-w64, I wasn't objecting, just pointing out to make sure the proper release flags are used.

Although, not using the inplace GCC has one potential issue: GHC has it's own runtime linker used for the interpreter (GHCi). The same runtime linker is used for Template Haskell code for runtime linking. Which means it has to understand the object files that GCC produces. 

Because we've always had the versions fixed we were less prone to problems arising by updating gcc/binutils. (Such problems happen every now and then on linux, where GCC isn't bundled with GHC). So by decoupling the two and allowing you to use any version of GCC things may break more often.

But again I am not objecting here. We've been upgrading the packaged compiler quite frequently lately. So issues would likely be quickly fixed.

**Comment by mati865** [comment]
@Mistuke, that explains a lot. Thank you!

File: mingw-w64-ghc/0001-Switch-to-LLVM-version-3.8.patch (hunk 3 above)
Comment (ncihnegn) [inline mingw-w64-ghc/0001-Switch-to-LLVM-version-3.8.patch]: I am not sure why, but my config.log shows _release='1'_.

**Comment by ncihnegn** [comment]
appveyor has errors like [this](https://github.com/Alexpux/MSYS2-packages/issues/690)
 >      3 [main] pacman (1752) C:\msys64\usr\bin\pacman.exe: *** fatal error - cygheap base mismatch detected - 0x180339400/0x1802FE408.
> This problem is probably due to using incompatible versions of the cygwin DLL.
> Search for cygwin1.dll using the Windows Start->Find/Search facility
> and delete all but the most recent version.  The most recent version *should*
> reside in x:\cygwin\bin, where 'x' is the drive on which you have
> installed the cygwin distribution.  Rebooting is also suggested if you
> are unable to find another cygwin DLL.
>       1 [main] pacman 2852 fork: child -1 - forked process 1752 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
> error: could not fork a new process (Resource temporarily unavailable)

Are we running Windows insider builds in appveyor?

**Comment by mati865** [comment]
I don't think there are Insider hosts on AppVeyor.

**Comment by hongy19** [comment]
hi all,
when I try to compile, i meet following error.
does anyone know how to solve it?

by the way, where could find a complied version of mingw-GHC? I don't find it in repository. 

'''
==> Starting prepare()...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 4653k  100 4653k    0     0   106k      0  0:00:43  0:00:43 --:--:--  153k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  147M  100  147M    0     0   119k      0  0:21:06  0:21:06 --:--:--  322k
Archive:  cabal-install-1.24.0.0-x86_64-unknown-mingw32.zip
  inflating: host/bin/cabal.exe
Downloading the latest package list from hackage.haskell.org
Skipping download: local and remote files match.
Resolving dependencies...
Configuring alex-3.2.1...
Configuring happy-1.19.5...
Building alex-3.2.1...
Building happy-1.19.5...
Failed to install happy-1.19.5
Build log ( C:\Users\ehongya\AppData\Roaming\cabal\logs\happy-1.19.5.log ):
Building happy-1.19.5...
Preprocessing executable 'happy' for happy-1.19.5...
[ 1 of 18] Compiling GenUtils         ( src\GenUtils.lhs, dist\build\happy\happy-tmp\GenUtils.o )
[ 2 of 18] Compiling Target           ( src\Target.lhs, dist\build\happy\happy-tmp\Target.o )
[ 3 of 18] Compiling AbsSyn           ( src\AbsSyn.lhs, dist\build\happy\happy-tmp\AbsSyn.o )
[ 4 of 18] Compiling ParamRules       ( src\ParamRules.hs, dist\build\happy\happy-tmp\ParamRules.o )

src\ParamRules.hs:4:1: error:
    Failed to load interface for ▒▒Control.Monad.Writer▒▒
    There are files missing in the ▒▒mtl-2.2.1▒▒ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.

src\ParamRules.hs:5:1: error:
    Failed to load interface for ▒▒Control.Monad.Error▒▒
    There are files missing in the ▒▒mtl-2.2.1▒▒ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.
cabal: Leaving directory 'C:\msys64\tmp\cabal-tmp-18185\happy-1.19.5'
Failed to install alex-3.2.1
Build log ( C:\Users\ehongya\AppData\Roaming\cabal\logs\alex-3.2.1.log ):
Building alex-3.2.1...
Preprocessing executable 'alex' for alex-3.2.1...
[ 1 of 22] Compiling Set              ( src\Set.hs, dist\build\alex\alex-tmp\Set.o )
[ 2 of 22] Compiling DFS              ( src\DFS.hs, dist\build\alex\alex-tmp\DFS.o )
[ 3 of 22] Compiling Sort             ( src\Sort.hs, dist\build\alex\alex-tmp\Sort.o )
[ 4 of 22] Compiling Data.Ranged.Boundaries ( src\Data\Ranged\Boundaries.hs, dist\build\alex\alex-tmp\Data\Ranged\Boundaries.o )

src\Data\Ranged\Boundaries.hs:24:1: error:
    Failed to load interface for ▒▒Test.QuickCheck▒▒
    There are files missing in the ▒▒QuickCheck-2.9.2▒▒ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.
cabal: Leaving directory 'C:\msys64\tmp\cabal-tmp-18184\alex-3.2.1'
cabal.exe: Error: some packages failed to install:
alex-3.2.1 failed during the building phase. The exception was:
ExitFailure 1
happy-1.19.5 failed during the building phase. The exception was:
ExitFailure 1
==> ERROR: A failure occurred in prepare().
    Aborting...
'''

**Comment by Mistuke** [comment]
Your package database is corrupt in your stage-0 compiler is corrupt.. do what is suggested, run `ghc-pkg check` and reinstall `QuickCheck`.

**Comment by hongy19** [comment]
@Mistuke thanks for your feedback. I have successed to compile ghc, but it has issue when running it.
ghc

do you know if there is any compiled ghc package that I could use/installed directly? I try to google it, but I couldn't find any.

Thanks

**Comment by ncihnegn** [comment]
You should be able to build this in a clean environment. Likely your cabal is messed up.

**Comment by Mistuke** [comment]
@hongy19 sorry, I forgot to respond. `_lock_file` is provided by the C runtime. If https://sourceforge.net/p/mingw/bugs/2304/ was actually merged for GCC 6.2 that would explain your breakage. I haven't tested 6.2 yet as the official release is still on 5.2. Which OS version are you running? 

this is the only GHC package for msys2. Official builds are here https://www.haskell.org/ghc/ as tarballs. However this version in msys2 has some other patches applied that will be in 8.0.2.

**Pull request merged**